lib/Parse/ParseDecl.cpp

```diff
@@ -1530,7 +1529,133 @@ bool Parser::parseNewDeclAttribute(DeclAttributes &Attributes, SourceLoc AtLoc,
                                                  /*Implicit=*/false));
     break;
   }
-
+  case DAK_OriginallyDefinedIn: {
+    auto LeftLoc = Tok.getLoc();
+    if (!consumeIf(tok::l_paren)) {
+      diagnose(Loc, diag::attr_expected_lparen, AttrName,
+               DeclAttribute::isDeclModifier(DK));
+      return false;
+    }
+    SourceLoc RightLoc;
+    enum class NextSegmentKind: uint8_t {
+      ModuleName = 0,
+      VersionNumber,
+      End,
+    };
+    NextSegmentKind NK = NextSegmentKind::ModuleName;
+    StringRef OriginalModuleName;
+    PlatformKind Platform = PlatformKind::none;
+    Optional<llvm::VersionTuple> MovedVersion;
+    StringRef AttrName = "@_originalDefinedIn";
+    if (parseList(tok::r_paren, LeftLoc, RightLoc, false,
+                  diag::originally_defined_in_missing_rparen,
+                  SyntaxKind::Unknown, [&]() -> ParserStatus {
+      SWIFT_DEFER {
+        if (NK != NextSegmentKind::End) {
+          NK = (NextSegmentKind)((uint8_t)NK + (uint8_t)1);
+        }
+      };
+      switch (NK) {
+      // Parse 'module: "original_module_name"'.
+      case NextSegmentKind::ModuleName: {
+        auto Label = Tok.getText();
+        auto LableLoc = Tok.getLoc();
+        auto expectModuleName = [&]() {
+          diagnose(LableLoc,
+                   diag::originally_defined_in_need_original_module_name);
+        };
+        // Parse "module:"
+        if (!consumeIf(tok::identifier)) {
+          expectModuleName();
+          return makeParserError();
+        }
+        if (Label != "module") {
+          expectModuleName();
+          return makeParserError();
+        }
+        if (!consumeIf(tok::colon)) {
+          expectModuleName();
+          return makeParserError();
+        }
+        // Parse the next string literal as the original module name.
+        auto ModuleNameLoc = Tok.getLoc();
+        if (Tok.is(tok::string_literal)) {
+          auto NameOp = getStringLiteralIfNotInterpolated(Tok.getLoc(),
+                                                          "original module name");
+          if (NameOp.hasValue())
+            OriginalModuleName = *NameOp;
+          consumeToken();
+        }
+        if (OriginalModuleName.empty()) {
+          diagnose(ModuleNameLoc,
+                   diag::originally_defined_in_need_nonempty_module_name);
+          return makeParserError();
+        }
+        return makeParserSuccess();
+      }
+      // Parse 'OSX 13.13'.
+      case NextSegmentKind::VersionNumber: {
+        if ((Tok.is(tok::identifier) || Tok.is(tok::oper_binary_spaced)) &&
```

Comment:
I don't think we accept `tok::oper_binary_spaced` as a module name.

Reply:
This is actually for the platform name being `*`. We support `*` as platform name in the availability attribute but not here.